Support for crostic puzzles

cathysax opened the pull request:
Added 'crostic' to puzzleModes, and new 'text-data-wrap' option for wrapping around side-to-side when typing or arrowing. Letter correlation between grid and clue is implemented with a new 'data-link-id' similar to 'data-extract-id' (which supports having extract cells in a crostic grid). New 'data-clue-locations' option of 'crostic' enables output similar to printed grids with numbers upper left and letters upper right. Implementation example is in reference-options.html.

cathysax committed:
Support for crostic puzzles

cathysax committed:
Support for crostic puzzles

tabascq commented on puzzlejs/puzzle.js:
```diff
@@ -2,7 +2,8 @@
  */
-var puzzleJsFolderPath = document.currentScript.src.replace("puzzle.js", "");
+//var puzzleJsFolderPath = document.currentScript.src.replace("puzzle.js", "");
+var puzzleJsFolderPath = document.currentScript.src.split("puzzle.js")[0];
```
Can you explain the scenario that made you want to change this?

cathysax replied:
I use the trick of doing "?v#" after a filename in order to force the browser to reload (when that # is changed) instead of using a stale, cached version. With the original code, that left "?v2" at the beginning of the path, which got pre-pended to the file it was trying to load (svg line stuff). My current line:
<script src="puzzlejs/puzzle.js?v4" language="JavaScript" type="text/javascript"></script>

tabascq replied:
That makes sense, thanks for the explanation!

tabascq commented on puzzlejs/puzzle.js:
```diff
@@ -354,7 +362,20 @@ function PuzzleEntry(p, index) {
 
             if (!td) {
-                return false;
+                if (!this.options["data-text-wrap"])
```
love this support - thanks!

cathysax committed:
Tabstop for each PuzzleGrid

Previously had tabstop only for first grid if multiple puzzle-grid DIVs in a puzzle-entry DIV.

tabascq commented on puzzlejs/puzzle.js:
```diff
@@ -2037,22 +2068,43 @@ function PuzzleGrid(puzzleEntry, options, container, puzzleId) {
-                          downClues[downClueIndex].setAttribute("data-down-cluenumber", clueIndicator);
-                          downClues[downClueIndex].setAttribute("value", clueIndicator);
-                          downClueIndex++;
+                        if (this.options["data-clue-locations"] == "crostic") {
```
There are some things that I like very much here:

- The description of linking IDs with multiple parts and knowing what to do with those parts
- The ability to simply move overlay components (clues, extracts, future things) around with no added CSS necessary

That said, I am not keen on merging the concepts of links (what I have called "extracts", but I think "links" is very very evocative and I am regretting my name choice!) and clues. I have already used both in the same cell in several puzzles (e.g. a crossword or KenKen where highlighted cells extract an answer) and feel that this would be limiting. 

I'd like to try this a slightly different way than what you've done here, which I believe preserves the spirit of what you're after:

- the ability to specify something like `data-extracts="C|1 A|2 C|3 C|4 C|5 B|6 C|7 B|8 B|9 B|10 A|11 C|12 A|13"`, where the linking codes are 1/2/3 etc but the displayed text is C1/A2/C3 etc. This is similar to what you're doing in data-clue-indicators but keeps the functionality in data-extracts and matches the layout that I see in crostic puzzles elsewhere.
- properties like `data-extract-position="top-right"`, which would add a `pos-top-right` class to all extracts (and have good defaults in the CSS). The CSS properties would also directly transfer to a `data-clue-position` property as well.
- a similar `avoid-top-right`, etc. class for cell contents, perhaps automatically set when there are clues or extracts.
- `data-modes` like `crostic-body` and `crostic-clue` that set sensible defaults from the above.

So it would look something like:

```
<div class="puzzle-entry" data-mode="crostic">
    <div style="min-width: 400px; background-color: white; padding: 10px;">
        <div class="puzzle-grid" data-mode="crostic-body" data-text="#####@#|#######"
            data-extracts="C|1 A|2 C|3 C|4 C|5 B|6 C|7 B|8 B|9 B|10 A|11 C|12 A|13"></div>
        <br>
        <div style="display: grid; grid-template-columns: 160px auto; row-gap: 20px; align-items: center;">
            <div>A. Not new</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="###" data-extracts="11 2 13"></div>
            <div>B. Left, on a map</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="####" data-extracts="10 6 8 9"></div>
            <div>C. Lights, Camera, ____!</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="######" data-extracts="7 1 5 3 12 4"></div>
        </div>
    </div>
</div>
```

I recognize that this is a lot and I am not asking you to do all of it, but I'd like to know if it covers all of the issues you're trying to address in a similarly straightforward way.

Does that get everything?

cathysax replied:
I had actually added links as a new concept, specifically so that I wasn't using extracts, leaving that still available. (I tested by created a crostic that also had extracts.) I think your version won't support having an extract of just some cells from the crostic. Am I missing something?

tabascq replied:
The issue is that it appears that you are using `data-clue-indicators` to hold things that are meant to be links, whereas I intend `data-extracts` to hold the links. [I agree that my property naming isn't the absolute best for carrying out this discussion, but semantically this is what I am after.] That said, perhaps *I* am missing something.

cathysax replied:
I really like the concept of being able to specify overlay location (data-clue-position, data-extract-position, etc) without needing to use CSS!

cathysax replied:
Having crostic-body and crostic-clue as separate types seems nice. Great that the different look for each of them could be automatic from those names instead of needing to set the layout details on each puzzle-grid div.

tabascq replied:
Thanks - and I am trying to specify it in a puzzle-type-agnostic way, so that puzzle types we've never even thought of can benefit.

cathysax replied:
> The issue is that it appears that you are using `data-clue-indicators` to hold things that are meant to be links, whereas I intend `data-extracts` to hold the links. [I agree that my property naming isn't the absolute best for carrying out this discussion, but semantically this is what I am after.] That said, perhaps _I_ am missing something.

Maybe using 'data-clue-indicators' isn't the best solution, but I was trying to be as minimally disruptive as possible. My goal was to have a way to indicate in two different puzzle-grid DIVs that a given '#' represented the same answer letter. I originally used data-extracts (from your crostic example), but I didn't like that it prevented using that feature for extracting a few cells from the grid for a final answer. (Not that I need that now, just thinking of future use.)

So, I chose to consider the data-clues as the identifier that should match across grids. I added the new data-link-id to accomplish that, so the extract feature is still possible for a crostic grid. (data-link-id works like data-extract-id to keep the cells in sync). 

Ultimately, if you don't like the idea of expanding clues to something that can be linked, it sounds like adding a new link-like parameter would be a solution.

One other comment related to crostics -- my personal preference is to leave the letter off for each of the individual clues, which is what necessitates having something that understands the concept of number+letter in the main grid with the letter essentially just being a decoration; the actual linking is based purely on the numbers (in both grids). That really feels to me like a clue-indicator, which is why I chose that implementation. That said, I'm fine with however you want to implement it!

tabascq replied:
Oh! Here is what I am getting wrong about this. I grew up solving puzzles such as crostics in GAMES Magazine, which used a different format than seems to be in use today!

In GAMES Magazine, the number and letter were fused together into a single item (though I checked my back issues and see that I got the order backwards; it's 1A, not A1). I see that in the NYT, New Yorker, and elsewhere, the format is consistently different from this, and that format is what I assume you are trying to replicate. I haven't done a lot of crostics in the last 20+ years and have not noticed the change in standard (or maybe GAMES was always nonstandard and I just never knew). I now 100% understand why you are saying that the number looks like a clue indicator (though I still don't want to think of it as a clue indicator). 

Here is a new version that I think will make everything work and also be super flexible; I've only changed the `crostic-body` element. I've explicitly written the `data-extract-position` attribute below to help explain, though this would really be folded into `data-mode="crostic"` and would not generally need to be specified.

```
<div class="puzzle-entry" data-mode="crostic">
    <div style="min-width: 400px; background-color: white; padding: 10px;">
        <div class="puzzle-grid" data-mode="crostic-body" data-text="#####@#|#######"
            data-extract-position="top-left|top-right"
            data-extracts="1|C 2|A 3|C 4|C 5|C 6|B 7|C 8|B 9|B 10|B 11|A 12|C 13|A"></div>
        <br>
        <div style="display: grid; grid-template-columns: 160px auto; row-gap: 20px; align-items: center;">
            <div>A. Not new</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="###" data-extracts="11 2 13"></div>
            <div>B. Left, on a map</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="####" data-extracts="10 6 8 9"></div>
            <div>C. Lights, Camera, ____!</div>
            <div class="puzzle-grid" data-mode="crostic-clue" data-text="######" data-extracts="7 1 5 3 12 4"></div>
        </div>
    </div>
</div>
```

This would split the extract code into two components and separately position them; it would put the numbers in the top left of each cell and the letters in the top right of each cell, but the placement could be changed for whatever purpose the author deemed fit; any number of parts in any location deemed desirable. A screenreader would read out all the parts concatenated together. The first term of the code (in this case, the number) would be the true extract code for linking purposes, and that choice could be straightforwardly extended further in the future if desired (I have an idea here but don't want to confuse things at present).

The default value of `data-extract-position` would be `bottom-right` to match current behavior in non-crostic scenarios. The `crostic-body` and `crostic-clue` modes would both overwrite that default.

cathysax replied:
That's getting closer to what I envision. I like the "|" notation to indicate multiple parts for the cell label and having the parallel 'position' param. (I'd been using '_' but '|' seems better.)
The issue is that it's still 'stealing' the extract feature, which means that we can't extract letters from the grid for a final answer.
I'm going to add a comment at the end of this page, as part of the general conversation (instead of on a specific piece of code) to start talking about the individual features and implementation ideas.

tabascq replied:
I think the multiple-levels-of-linking issue is going to be solvable with creative marking of extraction labels, especially if one of the values in `data-extract-position` is `none`. Then I can use a single underlying code across all three locations in your scenario, and display two different types of linking for two slightly different purposes, all in the same single property. I can work out a full example of this if it isn't clear.

Whether the base property name should be "extract" or "link" is a separate matter, of course. I'm not ready for the breaking change at present, but if I receive more feedback like yours I could get there.

cathysax replied:
Adding another extract value and using none is a great solution. (No need to change the name to link.)
Then my only other issue is that I don't want the yellow coloring for all the cells, but I can override that with CSS on my puzzle's page.

tabascq reviewed (commented):
Thanks for the PR! I've got a somewhat lengthy comment in here but I like the overall direction very much.

cathysax commented:
I hadn't intended to close this; I was rearranging branches and am afraid that had a bad effect on this PR. I think I successfully un-closed it. I can re-submit if needed. Please advise.

tabascq commented:
> I hadn't intended to close this; I was rearranging branches and am afraid that had a bad effect on this PR. I think I successfully un-closed it. I can re-submit if needed. Please advise.

No worries; still looks good to me!

cathysax commented:
One issue: I uploaded the tab stop code to this branch, so it's now included in the code here. I hadn't realized that the PR was a live link to the branch, not a snapshot of it when I made the PR.
I'll do a better job with branches and PR with single features in the future. Sorry I messed that up this time. Once we figure out how we want the crostics to work, you can close this (without taking any code) and I'll create appropriate branches for each of the features so that you can consider them individually.

tabascq commented:
Not a problem at all; feel free to use this branch for everything you want in your personal version, and we can indeed carve it up later. No reason you need to manage multiple branches of stuff when you have a singular target in mind on your end.

cathysax commented:
I'm thinking that it would be helpful to list the individual features I'm hoping to add/use. Then we can discuss thoughts on how each should be implemented. I've been able to refine these ideas as we've chatted, and I expect I'll keep improving my understanding of how to implement things in keeping with your vision.

1. Link cells in main grid and clue grid, keeping contents in sync.
2. Two-part label, with [first] part [optionally] used for linking. This would be in main grid only.
3. Specify locations for each part of the label.
4. Cell contents shifted down, slightly smaller font (because top of cell full with labels).
5. Support text flow in regular reading order: wrap side-to-side when typing or arrowing, don't change auto-advance to vertical (arrowing up/down, space, click).
6. Tabstop for each sub-grid.
7. Support adding "?v#" suffix to puzzle.js link in HTML.

I'll add individual comments on these to make it easier to quote and discuss separately. Everything is a suggestion, and I appreciate your feedback on how to make it better fit your model, both in design and naming. (I know all the options all start with "data-" just wanted to be a little less verbose in the comments below.)

cathysax commented:
_1 / cell linking_
I'd prefer not to use "extract" so that feature can still be used to highlight a few grid cells for a final answer.
I had thought of the crostic numbers as being "clue-indicators" but I'm ok with not piggy-backing on that feature. Maybe this means we should add another option like "[data-]labels." Ultimately, though, I don't see how my labels aren't clues. They behave like clues, getting assigned to non-@ grid locations.
**I think the concept of linking is a separate feature, independent of the labeling.**
Proposal: add a new Boolean option, e.g. "[data-]clue-link," that indicates that same-id clues should be linked together. I'd use that option (instead of my current poor design choice of using clue-locations="crostic") to indicate when to set td.data-link-id.
(If you don't like the idea of using clues for the crostic labels, we can add data-labels="#|X..." and data-label-link="true.")

cathysax commented:
_2 / two-part label_
I like your plan of using the "|" notation to support multi-part labels. Some options for implementation:
1. clue-locations=crostic (current implementation, not good choice)
2. clue-locations=all, clue-multi-part=true (new Bool option)
3. clue-locations=all, with multi-part always an option via "|"
4. clue-locations=all-multi-part (new value)
5. new [data-]labels="1|C 2|B ..." option, which always supports multiple parts

(If using clue-location option, clue-indicators would have list of numbers/labels)
I lean toward 2 -- "all" is a well-defined existing behavior, and a new option for parsing as multi-part seems nicer for backward compatibility than assuming that it's ok to split on "|". I'd like to use the existing clue mechanism since that will help avoid dup code, but I'm ok with instead adding a new "label" concept.

cathysax commented:
_3 / label positions_
I like your plan of having a new parallel "...-position" option to specify the clue/label positions.

cathysax commented:
_4 / cell content position_
I had implemented the smaller/shifted-down content just for crostics, but that's clearly not in the spirit of the "use for anything" concept. Instead of adding CSS using the .crostic class (which otherwise has no purpose), I should have added a new option (and corresponding class for CSS); maybe "data-text-position" for this?

cathysax commented:
_5 / text flow_
I haven't [yet] implemented the second part. I implemented the first part with a new "text-wrap" option, but maybe it would be more appropriate to have a new option like "text-advance-style" of "crossword" (default) and "horizontal" that would cover the two behaviors.

tabascq commented:
1+2. I want a scheme where everything can be linked together easily and indicators are autogenerated/placed with good defaults based on the puzzle type. I want extractions in puzzles like crosswords or here's-a-list-of-words to be the most straightforward cases. That said, I do get that I'm making it hard to get highlighting of cells in a crostic.

Maybe the answer is closest to your option 5, which I know is fairly close to what you've been saying all along: `data-extracts` makes links, and there's a new `data-links` property that also makes links, but without the highlighting. The `data-links` property would most likely link all cells all the time, since adding new character types into `data-text` is "expensive" (since spare characters are a valuable and non-renewable resource). Note that I still don't think these things are exactly "clues" since clues can be repeated without linking in, say, a KenKen where multiple cells can have identical clues but different contents.

As for multi-part, I like always supporting them since there's no real reason to declare in advance.

4. Yes, something like this is what I have been envisioning as well, though I'm not sure I want to call it "position". I've been thinking something like `data-text-avoid="top"` for crostics, so the word `top` can match the use of `top` in the link positioning info.

5. Big thumbs up on `data-text-advance-style`. I think that works very well.

tabascq commented:
Slight modification to this: I could get behind the concept of someday treating the *letter* portion as a clue, since that part could function like it does in a crossword (highlighting the clue text etc). But getting into the highlighting of the clue text etc. is a whole new layer on this discussion that doesn't need to be added right away.

cathysax commented:
1+2 Good point about KenKen clues -- and lots of other Nikoli-style grids, which I should have thought about since those are the kinds of puzzles I usually make! I'm now totally on board with not using the 'clue' constructs for crostics. I think that means either (a) using extracts and doing something to provide a way to control the yellow coloring, or (b) adding new link- options for the labels and their positions.

4 The use of  `data-text-avoid="top"` to match the label position `top` phrasing seems nice.

cathysax commented:
> treating the _letter_ portion as a clue

I'm not sure what this is referring to. Was this in response to something I requested, or thinking about other ways of linking that may be requested in the future? (e.g. highlighting the clue corresponding to the currently-selected cell in a crostic grid)

tabascq commented:
> I think that means either (a) using extracts and doing something to provide a way to control the yellow coloring, or (b) adding new link- options for the labels and their positions.

I'm thinking there are two properties:

- `data-extracts`: only applies to characters in `data-text` that are `#`, and sets the `extract` class (yellow formatting by default)
- `data-links`: applies to all cells and does not set the `extract` class

I now think these properties have to have nearly identical propagation logic, but the propagation keys need to be unique to whichever of these two scopes they're defined in. It'll be a little tricky to handle a propagation cascade, but worth it. Cascade example: type a letter in a crostic clue which propagates to the crostic body via `data-links`, and then if that body cell is an extract cell, propagates again elsewhere via `data-extracts`.

tabascq commented:
> > treating the _letter_ portion as a clue
> 
> I'm not sure what this is referring to. Was this in response to something I requested, or thinking about other ways of linking that may be requested in the future? (e.g. highlighting the clue corresponding to the currently-selected cell in a crostic grid)

Your e.g. is precisely the scenario I meant to describe. I'm fine leaving this as a future expansion thing.

cathysax commented:
I was working on writing up a summary and realized that there's still a little more I'd like to iron out with the labels/links design.
I like the idea of having the crostic number and letter together (vs in parallel lists); much easier to make sure the letters stay in sync. But I don't like calling that a 'link' since it's really just the first letter that's the link. 'label' feels like the right name for that construct. So, what about having `data-label-link` to specify linking?
I envision this as being similar to -position where you can specify the kind of link: content (as I'm doing now, like extract), highlight (which could be useful for future expansion to highlight all the "A" clue letters [in both grids]), and possibly some other purpose as-yet-unknown...
Using data-label-link would create a param `data-link-content-id` (or `data-link-highlight-id`) set to the value from the corresponding entry in the data-labels list. As with -position, multiple values could be separated with |.
E.g.
_grid:_
```
data-labels="1|C 2|A 3|C 4|C 5|C 6|B 7|C 8|B 9|B 10|B 11|A 12|C 13|A"
data-label-position="top-left|top-right"
data-label-link="content"
```
_clue:_
```
data-labels="11 2 13"
data-label-position="top-left"
data-label-link="content"
```
Also related to labels, it seems like it would be nice to provide a way to skip a label since it's automatically assumed that there's a label for every non-@ cell. Maybe "\_"? E.g. "1|C 2|\_ 3|C \_|C" I don't have a good example of why I'd want that, but it seemed worth considering.

tabascq commented:
I was thinking we would just use `data-links` since in the more common non-acrostic cases the single code is enough. But I was expecting there would eventually be another property to control which portion of a link is used as the actual code; something like `data-link-code-pattern`, whose default is `0`. Integers here are indices into `cellLinkText.split("|")`; parts can be ignored, reordered, etc. as necessary, and the default should be good enough for almost everything. If I have a three-part link for some reason and I want the first and last parts to be relevant, I could specify `data-link-code-pattern="02"`. Note that I don't think this property is needed right away, but this is how I'd address things if the default did indeed prove limiting.

This is the idea that I previously said I didn't want to describe yet since it might have confused things :-) But since almost everything else is getting traction, makes sense to bring it up now.

cathysax commented:
I figured it was good to think through the full design even if only part will be implemented now.
From a naming standpoint, I like data-**labels** and data-**label**-position. (The name data-**link**-position doesn't seem as informative; it's not really the link's position, and I'm still thinking that the split label is a display issue, separate from the linking of cells for content or highlighting.)
I think I understand how your data-link-code-pattern concept would fit with this: it would indicate which label part (from .split("|")) would be used for linking. One question, though: if you specify "02" (or any other pair) would there be separate linked groups, i.e. would a value in the '2' position link with the same value in a '0' position, or would they form different groups (i.e. different JS param name)?
It seems to me that there's also the question of _what_ the link is for -- sharing content, highlighting, etc. That's a functionality thing, so needs code support, not just CSS. Might be nice for the option name to be informative about that.
For name, how about `data-label-link-content`? That groups it as part of the data-label options, and indicates that we're linking content. For the potential future feature of linking cells for highlighting, there could be a `data-label-link-highlight` option.

cathysax commented:
Sorry, meant to suggest that as `data-label-link-text` since "text" is how you refer to content.

cathysax commented:
A further thought on syntax if you wanted to have separate link groups for your "02" case: `data-label-link-text="foo||bar"` could indicate that we set JS params `data-link-foo-id="[first label val]"` and `data-link-bar-id="[third label val]"`

tabascq commented:
I'm not liking `label` as much because I already have two other things that present as labels - clues and extracts. It doesn't _have_ to be `link` but I like `label` less.

[In some parallel universe, everything overlaid in a cell might be defined as labels, with further behavioral properties specifying how cells with those labels link to other cells/clues/whatever, but that's not what puzzle.js is doing so far. Maybe a light will go on for me at some point and I will want to undertake a breaking change of that scope, but I'm not there at present.]

My "02" example is some artificial scenario where it takes the 0th term and the 2nd term, and concatenates them together to form a composite code. So for a link description of `A|B|C`, the underlying link code would be `AC`. I have no idea whatsoever what puzzle would want that; I'm just demonstrating that if "0th term only" proves limiting someday, there is a simple syntax that can do both simplistic things (e.g. simply "1" instead of the default "0", to access the next term) and things that go far beyond.

cathysax commented:
> I already have two other things that present as labels

Got it! I hadn't thought of it that way. Using link makes perfect sense now.

cathysax commented:
Ok, consolidating into what I think the spec is based on what we've discussed. If you've got more than a few tweaks, we can move this to Google Doc or something to collaborate.

new options:
- data-links -- space-delimeted list of labels to place in grid; assigns values sequentially to all grid locations that aren't @ (as for clue-indicators); "|" separates individual values to be placed independently in the same cell
- data-link-position -- position of label in cell; values: top-left, top-right, bottom-left, bottom-right; when multiple values are specified for cells, use "|" between positions here (correlated with values in data-links); implemented via DIV w/ CSS class (pos-top-right, etc); what should happen if there are values with no position specified? (no display of that value?)
- (future: data-clue-position -- lower on my priority list to add this; shouldn't be too hard, but might be better to make a separate feature/commit)
- data-text-avoid-position -- top, bottom (suggested adding "position" to your name since "text-avoid" seemed too general)
- data-text-advance-style -- values: "crossword" (default) and "horizontal" -- for latter, wrap side to side and don't change flow to vertical (arrow, click, space)

new modes:
- crostic-grid -- data-link-position="top-left|top-right"
- crostic-clue -- data-link-position="top-left"
- both would have: data-text-avoid-position="top" data-text-advance-style="horizontal"
- HTML sets data-links [not part of mode], #|X for grid, # for clue

puzzle.js:
- support adding "?v#" suffix to puzzle.js link in HTML.
- tabstop for each sub-grid -- I know you're planning to explore whether my code is sufficient. I'll leave it as is for now.
- need to propagate letter across all cells involved in extract and link network (e.g. clue->grid via link, then grid->answer via extract)

tabascq commented:
This spec looks great modulo a few minor nits:

`data-link-position`: should also support at least the following as well: `"top"`, `"bottom"`, `""` (which does not display at all). If there are more labels provided than data-link-position contains, then I guess treat them as `""` for now because that's super easy. That said, I can also see an argument for concatenating them into the last-specified position, and it would be easy to change this later if need be.

`data-text-advance-style`: I don't think `horizontal` is sufficiently evocative. `wrap`? `flow`?

`crostic-clue`: I think the default should perhaps be `bottom` to more closely match what I see in existing crostic puzzles.

cathysax commented:
hey, just realized -- we can't have separate crostic-grid and crostic-clue data-modes because data-mode is a puzzle-entry property, but grid and clue are needed at puzzle-grid level. I think that means sticking with a single crostic mode and just having grid & clue specify their own link-position.

tabascq commented:
I'm 100% fine with allowing `puzzle-grid` to support `data-mode` as well, and obviously thought it already did! :-)

cathysax commented:
There's a more general issue with modes and options on puzzle-grid -- the move()/etc fns use 'this' for the puzzle-**entry**, not puzzle-**grid**. The only way to get the advance / wrap functionality is to put the options (data-text-advance-on-type, data-text-advance-style) on the puzzle-entry, either directly or with a mode.
So, I'll use a 'crostic' mode on the puzzle-entry div. I'll set that up for the main grid. Then I'm planning to just manually set (override) the data-link-position for the clues.
I'm thinking it would be bad to enable my code that supports modes on puzzle-grid unless/until it's documented which options are actually supported for puzzle-grid.

tabascq commented:
I don't think this is an issue with modes, since modes are just collections of properties.

Using `this.activeGrid` instead of just `this` would probably fix this property and therefore permit the original plan.

cathysax commented:
ok, cool. btw, this would be an issue for anyone using `data-text-advance-on-type` in a puzzle-grid, not just crostics. I updated to include `activeGrid` for all uses of that in PuzzleEntry code.

cathysax commented:
Regarding the position of the link in the clues: Having the number at the bottom makes sense if there are simple underscore blanks with numbers underneath (as I'm doing in the print version), but I don't think it looks as nice to have numbers at the bottom of a box, and I like the box as a nice fill-me-in visual for the web page. I like having the number upper left to match the grid.

tabascq commented:
Makes sense, thanks - someday I may make an `edge-style` that does underscores and a `bottom-outside` position, but there's no rush on that and I hear you that it might not actually be better.

cathysax commented:
closed since PR 41 contained new implementation
